File: rb/JikanRest_sdk.rb
# JikanRest SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'JikanRest_types'


class JikanRestSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = JikanRestUtility.new
    @_utility = utility

    config = JikanRestConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features in the resolved order (make_options puts an explicit array
    # order first, else defaults to test-first). Ordering matters: the `test`
    # feature installs the base mock transport and the transport features
    # (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
    # must be added before them to sit at the base of the chain.
    feature_opts = JikanRestHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      featureorder = VoxgigStruct.getpath(@options, "__derived__.featureorder")
      if featureorder.is_a?(Array)
        featureorder.each do |fname|
          fopts = JikanRestHelpers.to_map(feature_opts[fname])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, JikanRestFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    JikanRestUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = JikanRestHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = JikanRestHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = JikanRestHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = JikanRestSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    # make_fetch_def returns a (fetchdef, err) tuple; destructure it and
    # return just the fetchdef Hash (raising on error) so callers — including
    # direct(), which indexes fetchdef["url"] — receive a Hash, mirroring the
    # ts/py prepare().
    fetchdef, fd_err = utility.make_fetch_def.call(ctx)
    raise fd_err if fd_err

    fetchdef
  end

  # Raw endpoint access is operator-controllable, like every entity op.
  # Blocking it means denying BOTH the 'direct' and 'graphql' tokens, since
  # either one reaches the same endpoint.
  def direct(fetchargs = {})
    return op_denied("direct") unless op_allowed?("direct")

    raw_request(fetchargs)
  end

  # Is this raw-access op permitted by the SDK's allow.op option?
  def op_allowed?(op)
    allow_op = VoxgigStruct.getpath(@options, "allow.op")
    allow_op.is_a?(String) && allow_op.include?(op)
  end

  def op_denied(op)
    allow_op = VoxgigStruct.getpath(@options, "allow.op")
    {
      "ok" => false,
      "err" => JikanRestError.new(
        "#{op}_allow",
        "JikanRestSDK: #{op}: operation not allowed by" \
        " SDK option allow.op value: \"#{allow_op}\""),
    }
  end

  # Ungated request path shared by direct and graphql, each of which checks
  # its own allow.op token first. Separate, rather than a flag on fetchargs:
  # a caller-supplied marker would let anyone opt straight back out of the
  # gate by passing it.
  def raw_request(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue JikanRestError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = JikanRestHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = JikanRestHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end

  # Raw GraphQL access: the pressure valve that makes the generated surface's
  # deliberate omissions (per-call selection sets, typed filter builders,
  # batching, subscriptions) livable — the whole schema stays reachable.
  #
  # Thin wrapper over the same prepare/fetch path direct uses, with the one
  # thing raw direct cannot do for GraphQL: a GraphQL failure rides HTTP 200
  # as a top-level `errors` array, so status alone would report a failed
  # query as ok.
  #
  # NOTE: like direct, this bypasses the feature pipeline — no retry,
  # ratelimit or paging features apply.
  def graphql(query, variables = nil, ctrl = nil)
    return op_denied("graphql") unless op_allowed?("graphql")

    res = raw_request({
      "method" => "POST",
      "headers" => { "content-type" => "application/json" },
      "body" => { "query" => query, "variables" => variables || {} },
      "ctrl" => ctrl || {},
    })

    # Errors are read BEFORE any status check: a GraphQL parse or validation
    # failure comes back as HTTP 400 carrying the standard { errors: [...] }
    # body, and the raw path represents a non-2xx as ok:false with no err —
    # so returning early on status would discard the server's own
    # diagnostics, which are the only useful part of that response.
    errors = VoxgigStruct.getpath(res, "data.errors")

    if errors.is_a?(Array) && !errors.empty?
      first = errors[0].is_a?(Hash) ? errors[0] : {}
      msg = first["message"]
      msg = "graphql error" if msg.nil? || msg.to_s.empty?
      res["ok"] = false
      res["err"] = JikanRestError.new(
        "graphql_error", "JikanRestSDK: graphql: #{msg}")
      res["graphql"] = errors
    end

    res
  end


  # Canonical facade: client.Anime.list / client.Anime.load({ "id" => ... })
  def Anime(data = nil)
    require_relative 'entity/anime_entity'
    AnimeEntity.new(self, data)
  end


  # Canonical facade: client.Character.list / client.Character.load({ "id" => ... })
  def Character(data = nil)
    require_relative 'entity/character_entity'
    CharacterEntity.new(self, data)
  end


  # Canonical facade: client.Club.list / client.Club.load({ "id" => ... })
  def Club(data = nil)
    require_relative 'entity/club_entity'
    ClubEntity.new(self, data)
  end


  # Canonical facade: client.External.list / client.External.load({ "id" => ... })
  def External(data = nil)
    require_relative 'entity/external_entity'
    ExternalEntity.new(self, data)
  end


  # Canonical facade: client.Genre.list / client.Genre.load({ "id" => ... })
  def Genre(data = nil)
    require_relative 'entity/genre_entity'
    GenreEntity.new(self, data)
  end


  # Canonical facade: client.Magazine.list / client.Magazine.load({ "id" => ... })
  def Magazine(data = nil)
    require_relative 'entity/magazine_entity'
    MagazineEntity.new(self, data)
  end


  # Canonical facade: client.Manga.list / client.Manga.load({ "id" => ... })
  def Manga(data = nil)
    require_relative 'entity/manga_entity'
    MangaEntity.new(self, data)
  end


  # Canonical facade: client.PeopleSearch.list / client.PeopleSearch.load({ "id" => ... })
  def PeopleSearch(data = nil)
    require_relative 'entity/people_search_entity'
    PeopleSearchEntity.new(self, data)
  end


  # Canonical facade: client.Person.list / client.Person.load({ "id" => ... })
  def Person(data = nil)
    require_relative 'entity/person_entity'
    PersonEntity.new(self, data)
  end


  # Canonical facade: client.Producer.list / client.Producer.load({ "id" => ... })
  def Producer(data = nil)
    require_relative 'entity/producer_entity'
    ProducerEntity.new(self, data)
  end


  # Canonical facade: client.Random.list / client.Random.load({ "id" => ... })
  def Random(data = nil)
    require_relative 'entity/random_entity'
    RandomEntity.new(self, data)
  end


  # Canonical facade: client.Recommendation.list / client.Recommendation.load({ "id" => ... })
  def Recommendation(data = nil)
    require_relative 'entity/recommendation_entity'
    RecommendationEntity.new(self, data)
  end


  # Canonical facade: client.Review.list / client.Review.load({ "id" => ... })
  def Review(data = nil)
    require_relative 'entity/review_entity'
    ReviewEntity.new(self, data)
  end


  # Canonical facade: client.Schedule.list / client.Schedule.load({ "id" => ... })
  def Schedule(data = nil)
    require_relative 'entity/schedule_entity'
    ScheduleEntity.new(self, data)
  end


  # Canonical facade: client.Season.list / client.Season.load({ "id" => ... })
  def Season(data = nil)
    require_relative 'entity/season_entity'
    SeasonEntity.new(self, data)
  end


  # Canonical facade: client.Top.list / client.Top.load({ "id" => ... })
  def Top(data = nil)
    require_relative 'entity/top_entity'
    TopEntity.new(self, data)
  end


  # Canonical facade: client.User.list / client.User.load({ "id" => ... })
  def User(data = nil)
    require_relative 'entity/user_entity'
    UserEntity.new(self, data)
  end


  # Canonical facade: client.UserAbout.list / client.UserAbout.load({ "id" => ... })
  def UserAbout(data = nil)
    require_relative 'entity/user_about_entity'
    UserAboutEntity.new(self, data)
  end


  # Canonical facade: client.UserClub.list / client.UserClub.load({ "id" => ... })
  def UserClub(data = nil)
    require_relative 'entity/user_club_entity'
    UserClubEntity.new(self, data)
  end


  # Canonical facade: client.UserFriend.list / client.UserFriend.load({ "id" => ... })
  def UserFriend(data = nil)
    require_relative 'entity/user_friend_entity'
    UserFriendEntity.new(self, data)
  end


  # Canonical facade: client.UserHistory.list / client.UserHistory.load({ "id" => ... })
  def UserHistory(data = nil)
    require_relative 'entity/user_history_entity'
    UserHistoryEntity.new(self, data)
  end


  # Canonical facade: client.UserStatistic.list / client.UserStatistic.load({ "id" => ... })
  def UserStatistic(data = nil)
    require_relative 'entity/user_statistic_entity'
    UserStatisticEntity.new(self, data)
  end


  # Canonical facade: client.UserUpdate.list / client.UserUpdate.load({ "id" => ... })
  def UserUpdate(data = nil)
    require_relative 'entity/user_update_entity'
    UserUpdateEntity.new(self, data)
  end


  # Canonical facade: client.WatchEpisode.list / client.WatchEpisode.load({ "id" => ... })
  def WatchEpisode(data = nil)
    require_relative 'entity/watch_episode_entity'
    WatchEpisodeEntity.new(self, data)
  end


  # Canonical facade: client.WatchPromo.list / client.WatchPromo.load({ "id" => ... })
  def WatchPromo(data = nil)
    require_relative 'entity/watch_promo_entity'
    WatchPromoEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = JikanRestSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: py/jikanrest_sdk.py
# JikanRest SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import JikanRestUtility
from core.spec import JikanRestSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import JikanRestBaseFeature
from features import _make_feature


class JikanRestSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = JikanRestUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return JikanRestUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = JikanRestSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def anime(self):
        """Idiomatic facade: client.anime.list() / client.anime.load({"id": ...})."""
        from entity.anime_entity import AnimeEntity
        cached = getattr(self, "_anime", None)
        if cached is None:
            cached = AnimeEntity(self, None)
            self._anime = cached
        return cached

    def Anime(self, data=None):
        # Deprecated: use client.anime instead.
        from entity.anime_entity import AnimeEntity
        return AnimeEntity(self, data)


    @property
    def character(self):
        """Idiomatic facade: client.character.list() / client.character.load({"id": ...})."""
        from entity.character_entity import CharacterEntity
        cached = getattr(self, "_character", None)
        if cached is None:
            cached = CharacterEntity(self, None)
            self._character = cached
        return cached

    def Character(self, data=None):
        # Deprecated: use client.character instead.
        from entity.character_entity import CharacterEntity
        return CharacterEntity(self, data)


    @property
    def club(self):
        """Idiomatic facade: client.club.list() / client.club.load({"id": ...})."""
        from entity.club_entity import ClubEntity
        cached = getattr(self, "_club", None)
        if cached is None:
            cached = ClubEntity(self, None)
            self._club = cached
        return cached

    def Club(self, data=None):
        # Deprecated: use client.club instead.
        from entity.club_entity import ClubEntity
        return ClubEntity(self, data)


    @property
    def external(self):
        """Idiomatic facade: client.external.list() / client.external.load({"id": ...})."""
        from entity.external_entity import ExternalEntity
        cached = getattr(self, "_external", None)
        if cached is None:
            cached = ExternalEntity(self, None)
            self._external = cached
        return cached

    def External(self, data=None):
        # Deprecated: use client.external instead.
        from entity.external_entity import ExternalEntity
        return ExternalEntity(self, data)


    @property
    def genre(self):
        """Idiomatic facade: client.genre.list() / client.genre.load({"id": ...})."""
        from entity.genre_entity import GenreEntity
        cached = getattr(self, "_genre", None)
        if cached is None:
            cached = GenreEntity(self, None)
            self._genre = cached
        return cached

    def Genre(self, data=None):
        # Deprecated: use client.genre instead.
        from entity.genre_entity import GenreEntity
        return GenreEntity(self, data)


    @property
    def magazine(self):
        """Idiomatic facade: client.magazine.list() / client.magazine.load({"id": ...})."""
        from entity.magazine_entity import MagazineEntity
        cached = getattr(self, "_magazine", None)
        if cached is None:
            cached = MagazineEntity(self, None)
            self._magazine = cached
        return cached

    def Magazine(self, data=None):
        # Deprecated: use client.magazine instead.
        from entity.magazine_entity import MagazineEntity
        return MagazineEntity(self, data)


    @property
    def manga(self):
        """Idiomatic facade: client.manga.list() / client.manga.load({"id": ...})."""
        from entity.manga_entity import MangaEntity
        cached = getattr(self, "_manga", None)
        if cached is None:
            cached = MangaEntity(self, None)
            self._manga = cached
        return cached

    def Manga(self, data=None):
        # Deprecated: use client.manga instead.
        from entity.manga_entity import MangaEntity
        return MangaEntity(self, data)


    @property
    def people_search(self):
        """Idiomatic facade: client.people_search.list() / client.people_search.load({"id": ...})."""
        from entity.people_search_entity import PeopleSearchEntity
        cached = getattr(self, "_people_search", None)
        if cached is None:
            cached = PeopleSearchEntity(self, None)
            self._people_search = cached
        return cached

    def PeopleSearch(self, data=None):
        # Deprecated: use client.people_search instead.
        from entity.people_search_entity import PeopleSearchEntity
        return PeopleSearchEntity(self, data)


    @property
    def person(self):
        """Idiomatic facade: client.person.list() / client.person.load({"id": ...})."""
        from entity.person_entity import PersonEntity
        cached = getattr(self, "_person", None)
        if cached is None:
            cached = PersonEntity(self, None)
            self._person = cached
        return cached

    def Person(self, data=None):
        # Deprecated: use client.person instead.
        from entity.person_entity import PersonEntity
        return PersonEntity(self, data)


    @property
    def producer(self):
        """Idiomatic facade: client.producer.list() / client.producer.load({"id": ...})."""
        from entity.producer_entity import ProducerEntity
        cached = getattr(self, "_producer", None)
        if cached is None:
            cached = ProducerEntity(self, None)
            self._producer = cached
        return cached

    def Producer(self, data=None):
        # Deprecated: use client.producer instead.
        from entity.producer_entity import ProducerEntity
        return ProducerEntity(self, data)


    @property
    def random(self):
        """Idiomatic facade: client.random.list() / client.random.load({"id": ...})."""
        from entity.random_entity import RandomEntity
        cached = getattr(self, "_random", None)
        if cached is None:
            cached = RandomEntity(self, None)
            self._random = cached
        return cached

    def Random(self, data=None):
        # Deprecated: use client.random instead.
        from entity.random_entity import RandomEntity
        return RandomEntity(self, data)


    @property
    def recommendation(self):
        """Idiomatic facade: client.recommendation.list() / client.recommendation.load({"id": ...})."""
        from entity.recommendation_entity import RecommendationEntity
        cached = getattr(self, "_recommendation", None)
        if cached is None:
            cached = RecommendationEntity(self, None)
            self._recommendation = cached
        return cached

    def Recommendation(self, data=None):
        # Deprecated: use client.recommendation instead.
        from entity.recommendation_entity import RecommendationEntity
        return RecommendationEntity(self, data)


    @property
    def review(self):
        """Idiomatic facade: client.review.list() / client.review.load({"id": ...})."""
        from entity.review_entity import ReviewEntity
        cached = getattr(self, "_review", None)
        if cached is None:
            cached = ReviewEntity(self, None)
            self._review = cached
        return cached

    def Review(self, data=None):
        # Deprecated: use client.review instead.
        from entity.review_entity import ReviewEntity
        return ReviewEntity(self, data)


    @property
    def schedule(self):
        """Idiomatic facade: client.schedule.list() / client.schedule.load({"id": ...})."""
        from entity.schedule_entity import ScheduleEntity
        cached = getattr(self, "_schedule", None)
        if cached is None:
            cached = ScheduleEntity(self, None)
            self._schedule = cached
        return cached

    def Schedule(self, data=None):
        # Deprecated: use client.schedule instead.
        from entity.schedule_entity import ScheduleEntity
        return ScheduleEntity(self, data)


    @property
    def season(self):
        """Idiomatic facade: client.season.list() / client.season.load({"id": ...})."""
        from entity.season_entity import SeasonEntity
        cached = getattr(self, "_season", None)
        if cached is None:
            cached = SeasonEntity(self, None)
            self._season = cached
        return cached

    def Season(self, data=None):
        # Deprecated: use client.season instead.
        from entity.season_entity import SeasonEntity
        return SeasonEntity(self, data)


    @property
    def top(self):
        """Idiomatic facade: client.top.list() / client.top.load({"id": ...})."""
        from entity.top_entity import TopEntity
        cached = getattr(self, "_top", None)
        if cached is None:
            cached = TopEntity(self, None)
            self._top = cached
        return cached

    def Top(self, data=None):
        # Deprecated: use client.top instead.
        from entity.top_entity import TopEntity
        return TopEntity(self, data)


    @property
    def user(self):
        """Idiomatic facade: client.user.list() / client.user.load({"id": ...})."""
        from entity.user_entity import UserEntity
        cached = getattr(self, "_user", None)
        if cached is None:
            cached = UserEntity(self, None)
            self._user = cached
        return cached

    def User(self, data=None):
        # Deprecated: use client.user instead.
        from entity.user_entity import UserEntity
        return UserEntity(self, data)


    @property
    def user_about(self):
        """Idiomatic facade: client.user_about.list() / client.user_about.load({"id": ...})."""
        from entity.user_about_entity import UserAboutEntity
        cached = getattr(self, "_user_about", None)
        if cached is None:
            cached = UserAboutEntity(self, None)
            self._user_about = cached
        return cached

    def UserAbout(self, data=None):
        # Deprecated: use client.user_about instead.
        from entity.user_about_entity import UserAboutEntity
        return UserAboutEntity(self, data)


    @property
    def user_club(self):
        """Idiomatic facade: client.user_club.list() / client.user_club.load({"id": ...})."""
        from entity.user_club_entity import UserClubEntity
        cached = getattr(self, "_user_club", None)
        if cached is None:
            cached = UserClubEntity(self, None)
            self._user_club = cached
        return cached

    def UserClub(self, data=None):
        # Deprecated: use client.user_club instead.
        from entity.user_club_entity import UserClubEntity
        return UserClubEntity(self, data)


    @property
    def user_friend(self):
        """Idiomatic facade: client.user_friend.list() / client.user_friend.load({"id": ...})."""
        from entity.user_friend_entity import UserFriendEntity
        cached = getattr(self, "_user_friend", None)
        if cached is None:
            cached = UserFriendEntity(self, None)
            self._user_friend = cached
        return cached

    def UserFriend(self, data=None):
        # Deprecated: use client.user_friend instead.
        from entity.user_friend_entity import UserFriendEntity
        return UserFriendEntity(self, data)


    @property
    def user_history(self):
        """Idiomatic facade: client.user_history.list() / client.user_history.load({"id": ...})."""
        from entity.user_history_entity import UserHistoryEntity
        cached = getattr(self, "_user_history", None)
        if cached is None:
            cached = UserHistoryEntity(self, None)
            self._user_history = cached
        return cached

    def UserHistory(self, data=None):
        # Deprecated: use client.user_history instead.
        from entity.user_history_entity import UserHistoryEntity
        return UserHistoryEntity(self, data)


    @property
    def user_statistic(self):
        """Idiomatic facade: client.user_statistic.list() / client.user_statistic.load({"id": ...})."""
        from entity.user_statistic_entity import UserStatisticEntity
        cached = getattr(self, "_user_statistic", None)
        if cached is None:
            cached = UserStatisticEntity(self, None)
            self._user_statistic = cached
        return cached

    def UserStatistic(self, data=None):
        # Deprecated: use client.user_statistic instead.
        from entity.user_statistic_entity import UserStatisticEntity
        return UserStatisticEntity(self, data)


    @property
    def user_update(self):
        """Idiomatic facade: client.user_update.list() / client.user_update.load({"id": ...})."""
        from entity.user_update_entity import UserUpdateEntity
        cached = getattr(self, "_user_update", None)
        if cached is None:
            cached = UserUpdateEntity(self, None)
            self._user_update = cached
        return cached

    def UserUpdate(self, data=None):
        # Deprecated: use client.user_update instead.
        from entity.user_update_entity import UserUpdateEntity
        return UserUpdateEntity(self, data)


    @property
    def watch_episode(self):
        """Idiomatic facade: client.watch_episode.list() / client.watch_episode.load({"id": ...})."""
        from entity.watch_episode_entity import WatchEpisodeEntity
        cached = getattr(self, "_watch_episode", None)
        if cached is None:
            cached = WatchEpisodeEntity(self, None)
            self._watch_episode = cached
        return cached

    def WatchEpisode(self, data=None):
        # Deprecated: use client.watch_episode instead.
        from entity.watch_episode_entity import WatchEpisodeEntity
        return WatchEpisodeEntity(self, data)


    @property
    def watch_promo(self):
        """Idiomatic facade: client.watch_promo.list() / client.watch_promo.load({"id": ...})."""
        from entity.watch_promo_entity import WatchPromoEntity
        cached = getattr(self, "_watch_promo", None)
        if cached is None:
            cached = WatchPromoEntity(self, None)
            self._watch_promo = cached
        return cached

    def WatchPromo(self, data=None):
        # Deprecated: use client.watch_promo instead.
        from entity.watch_promo_entity import WatchPromoEntity
        return WatchPromoEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk
